app/models/mortality_tracking.rb

```diff
@@ -43,4 +43,38 @@ class MortalityTracking < ApplicationRecord
     COMMENTS: "Comments"
   }
 
+  def self.create_from_csv_data(attrs)
+    attrs['shl_case_number']                     = attrs.delete('shl_number')
+    attrs['number_morts']                     = attrs.delete('_morts')
+    new(attrs)
+  end
+
+  validates_inclusion_of :raw, in: [true, false]
+
+  validates_presence_of :mortality_date
+  # mortality_date has 'Unknown' and 'unkown' in csv
+  validates_presence_of :cohort
+  validates_presence_of :shl_case_number
+  validates_presence_of :spawning_date
+  validates_presence_of :number_morts
+
+  validates :number_morts, numericality: true, allow_blank: true
+  # number_morts has been '9+' and 'TBD' in csv
+
+  validates :processed_by_shl, format: { with: /Y|N/ }, allow_blank: true
+
+  def mortality_date=(mortality_date_str)
+    return unless mortality_date_str
+    write_attribute(:mortality_date, DateTime.strptime(mortality_date_str, '%m/%d/%y'))
+  rescue
+    write_attribute(:mortality_date, nil)
+  end
+
+  def spawning_date=(spawning_date_str)
+    return unless spawning_date_str
+    write_attribute(:spawning_date, DateTime.strptime(spawning_date_str, '%m/%d/%y'))
+  rescue
+    write_attribute(:spawning_date, nil)
```

Comment:
I'm not sure how I feel about this rescue here - if DateTime fails to parse it's input, then we'll fail silently here. Should we maybe surface this error to the user somehow?

Reply:
Good point. To be honest, I don't think this conversion of string to date should be in this method. I sense this is going beyond the responsibilities of the model and taking the role of converting a string to date behind the scenes. Instead, I'd put it in the `create_from_csv_data`  (even better in service object IMO) method and let it raise an error if there is an issue.

I'm going to make some modifications and see if it would fit. Though it might not work well for the changes in https://github.com/rubyforgood/abalone/pull/113 since it does not rescue.

Reply:
@edwinthinks my PR got merged, and it contains my solution to this issue you're having. I agree that these code should be moved out of the `model` object.

Where I work, we use an `interactor` object for these types of stuff. It goes: `controller` -> `interactor` -> `model`

In `interactor` is where we will do all the validation, and data manipulation/business rules if needed.

But for now, there's not a lot of business logic going on, and they're pretty basic, maybe in the future we can think of refactoring it out into another object.